[WIP] Add checks for particular compound distributions

akash9712 opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234". See
https://github.com/blog/1506-closing-issues-via-pull-requests .-->


#### Brief description of what is fixed or changed    
Add function to check if compounding particular distribution returns RVs
whose distribution is known already. As of now, this is only done for
univariate normal distribution with a normal mean.
Example:
```
>>> N1 = Normal('N1', 0, 1)
>>> N2 = Normal('N2', N1, 2)
>>> pspace(N2)
SingleContinuousPSpace(N2, NormalDistribution(0, sqrt(5)))
```
#### Other comments

akash9712 committed:
Compound distributions' PDF

PDF of distributions with random variable(s) as their parameters can now
be calculated. This is done by marginalizing over the latent
distribution.

akash9712 committed:
Added tests for compound random variable's PDF.

akash9712 committed:
Merge branch 'master' of https://github.com/sympy/sympy into compound_dist

akash9712 committed:
Revert changes to `pdf` in crv_types.py

akash9712 committed:
[WIP] Compound distributions return `...Handmadedistribution` objects

akash9712 committed:
Merge branch 'master' of https://github.com/sympy/sympy into compound_dist

akash9712 committed:
[Do not Merge] Add checks for particular compound distributions

Add function to check if compounding particular distribution returns RVs
whose distribution is known already. As of now, this is only done for
univariate normal distribution with a normal mean.
Example:
```
>>> N1 = Normal('N1', 0, 1)
>>> N2 = Normal('N2', N1, 2)
>>> pspace(N2)
SingleContinuousPSpace(N2, NormalDistribution(0, sqrt(5)))
```

Upabjojr commented on sympy/stats/crv.py:
```diff
@@ -138,12 +139,7 @@ def set(self):
                 "Set of Conditional Domain not Implemented")
 
 
-class ContinuousDistribution(Basic):
```
are you removing the distinction between continuous and discrete distributions?

akash9712 replied:
The difference is still present in other classes such as `SingleContinuousDistribution`, but `ContinuousDistribution` and `DiscreteDistribution` classes were identical, `__call__` being the only method described. Adding `handle_compound_dist` would also lead to some more code repitition.

Upabjojr commented on sympy/stats/crv_types.py:
```diff
@@ -845,7 +851,7 @@ def check(rate):
 
     def pdf(self, x):
-        return self.rate * exp(-self.rate*x)
+        return  self.rate * exp(-self.rate*x)
```
two spaces?

akash9712 replied:
Sorry, made a typo.

Upabjojr commented on sympy/stats/crv_types.py:
```diff
@@ -1948,7 +1953,7 @@ def Normal(name, mean, std):
 
     mu : Real number or a list representing the mean or the mean vector
     sigma : Real number or a positive definite sqaure matrix,
-         :math:`\sigma^2 > 0` the variance
```
sigma squares is the variance

Upabjojr commented on sympy/stats/rv.py:
```diff
@@ -121,6 +121,27 @@ def as_boolean(self):
         return And(self.fulldomain.as_boolean(), self.condition)
 
 
+class RandomDistribution(Basic):
```
what about `ProbabilityDistribution`?

Upabjojr commented on sympy/stats/rv.py:
```diff
@@ -121,6 +121,27 @@ def as_boolean(self):
+                lpdf = density(arg)(sym)
+                expr = expr.replace(arg, sym)*lpdf
+                if arg.pspace.is_Continuous:
+                    expr = integrate(expr, (sym, dom))
```
`integrate` does not currently support the set notation, but it should in the future. Better leave a note.

Upabjojr commented on sympy/stats/rv.py:
```diff
@@ -121,6 +121,27 @@ def as_boolean(self):
+                if arg.pspace.is_Continuous:
+                    expr = integrate(expr, (sym, dom))
+                elif arg.pspace.is_Discrete:
+                    if dom in (S.Integers, S.Naturals, S.Naturals0):
```
this should be transferred to `summation` or `Sum`. But better open an issue an leave it for another PR.

Upabjojr commented on sympy/stats/tests/test_continuous_rv.py:
```diff
@@ -20,7 +20,7 @@
 
 from sympy.stats.crv_types import NormalDistribution
-from sympy.stats.rv import IndependentProductPSpace
+from sympy.stats.crv import  SingleContinuousPSpace
```
two spaces

Upabjojr commented:
This PR looks good. It needs some more tests.

akash9712 committed:
Minor changes in stats.

Upabjojr commented:
```python
def test_random_parameters_given():
    mu = Normal('mu', 2, 3)
    meas = Normal('T', mu, 1)
    assert given(meas, Eq(mu, 5)) == Normal('T', 5, 1)
```
This test makes sense, we need to keep this behaviour.

Upabjojr commented:
```python
def test_conjugate_priors():
    mu = Normal('mu', 2, 3)
    x = Normal('x', mu, 1)
    assert isinstance(simplify(density(mu, Eq(x, y), evaluate=False)(z)),
            Integral)
```
In this case, the integral gets calculated. Do you prefer to change the test or shall we keep the integral unevaluated?

Upabjojr commented on sympy/stats/crv_types.py:
```diff
@@ -143,6 +143,15 @@ def rv(symbol, cls, args):
+        if  new_dist != None:
+            return rv(symbol, new_dist[0], new_dist[1])
+        _pdf = dist.pdf(symbol)
+        _pdf, _set = dist.handle_compound_dist(_pdf), dist.set
```
I would not calculate the integral at this level. I would recommend creating a marginalized joint distribution, without performing the integration.

Upabjojr commented:
We have `JointDistribution`. I think we need something called `MarginalDistribution` as well... the pspace of a compound distribution could then be `(X, MarginalDistribution(JointDistribution(...), ... ))`.

What do you think?

akash9712 commented:
> We have `JointDistribution`. I think we need something called `MarginalDistribution` as well... the pspace of a compound distribution could then be `(X, MarginalDistribution(JointDistribution(...), ... ))`

I think this could work, and would help if a condition is imposed on the latent distribution.

akash9712 committed:
[WIP] Added class `MarginalDistribution`

Added class `MarginalDistribution` that takes as input expression with
multiple random variables and a set of set random variables `rvs`.
Random variables other than `rvs` are marginalised out when calculating
the PDF.

akash9712 commented:
@Upabjojr I have created the class `MarginalDistribution`, but I am still working on how to fix the failing tests; I am trying to modify `given` so that the conditions on latent distributions can be imposed.

ArifAhmed1995 commented on sympy/stats/crv_types.py:
```diff
@@ -143,7 +144,13 @@ def rv(symbol, cls, args):
+    pspace = SingleContinuousPSpace(symbol, dist)
+    latent_dist = [arg for arg in args if isinstance(arg, RandomSymbol)]
+    _pdf = pspace.pdf
+    if latent_dist != []:
```
I guess a more pythonic way is to use `if latent_dist:`

ArifAhmed1995 commented on sympy/stats/drv_types.py:
```diff
@@ -15,7 +16,13 @@ def rv(symbol, cls, *args):
+    pspace = SingleDiscretePSpace(symbol, dist)
+    latent_dist = [arg for arg in args if isinstance(arg, RandomSymbol)]
+    _pdf = pspace.pdf
+    if latent_dist != []:
```
Same here...

ArifAhmed1995 commented on sympy/stats/joint_rv.py:
```diff
@@ -163,3 +165,45 @@ def marginal_distribution(rv, *indices):
+        return ProductSet((i.pspace.set for i in rvs))
+
+    def pdf(self, x):
+        expr = self.args[1]
```
It might be a good idea to extract all the used arguments into variables with proper names. Apart from `args[1]` which refers to the expression, I cannot make out what the other two refer to.
In future, it makes it difficult for other contributors to work and modify the code.

ArifAhmed1995 commented on sympy/stats/rv.py:
```diff
@@ -517,6 +521,8 @@ def pspace(expr):
     True
     """
     expr = sympify(expr)
+    if isinstance(expr, RandomSymbol) and expr.pspace != None:
```
Try to use `expr.pspace is not None`.

akash9712 commented:
@ArifAhmed1995 The work done in this PR was included in #14989. Closing this now.
